src/metabase/util/ssh.clj

```diff
@@ -1,32 +1,63 @@
+           org.apache.sshd.common.util.security.SecurityUtils
+           org.apache.sshd.server.forward.AcceptAllForwardingFilter))
 
 (def ^:private default-ssh-timeout 30000)
 
+(def ^:private client
+  (doto (SshClient/setUpDefaultClient)
+    (.start)
+    (.setForwardingFilter AcceptAllForwardingFilter/INSTANCE)))
+
 (defn start-ssh-tunnel
   "Opens a new ssh tunnel and returns the connection along with the dynamically
-   assigned tunnel entrance port. It's the callers responsibility to call .disconnect
+   assigned tunnel entrance port. It's the callers responsibility to call `close-tunnel`
    on the returned connection object."
-  [{:keys [tunnel-host tunnel-port tunnel-user tunnel-pass host port]}]
-  (let [connection (doto ^com.jcraft.jsch.Session (.getSession (new com.jcraft.jsch.JSch)
-                                                               ^String tunnel-user
-                                                               ^String tunnel-host
-                                                               tunnel-port)
-                     (.setPassword ^String tunnel-pass)
-                     (.setConfig "StrictHostKeyChecking" "no")
-                     (.connect default-ssh-timeout)
-                     (.setPortForwardingL 0 host port))
-        input-port (some-> (.getPortForwardingL connection)
-                           first
-                           (str/split #":")
-                           first
-                           (Integer/parseInt))]
-    (assert (number? input-port))
-    (log/info (u/format-color 'cyan "creating ssh tunnel %s@%s:%s -L %s:%s:%s" tunnel-user tunnel-host tunnel-port input-port host port))
-    [connection input-port]))
+  [{:keys [tunnel-host tunnel-port tunnel-user tunnel-pass tunnel-private-key tunnel-private-key-passphrase host port]}]
+  (let [conn-future (.connect ^SshClient client ^String tunnel-user ^String tunnel-host ^Integer tunnel-port)
+        conn-status (.verify ^ConnectFuture conn-future default-ssh-timeout)
+        session (.getSession ^SessionHolder conn-status)
+
+        _ (when tunnel-pass
+            (.addPasswordIdentity ^ClientSession session tunnel-pass))
+        _ (when tunnel-private-key
+            (let [resource-key (proxy [AbstractIoResource] [(class "key") "key"])
+                  password-provider (proxy [FilePasswordProvider] []
+                                      (getPassword [_ _ _]
+                                        tunnel-private-key-passphrase)
+                                      (handleDecodeAttemptResult [_ _ _ _ _]
+                                        FilePasswordProvider$ResourceDecodeResult/TERMINATE))
+
+                  ids (SecurityUtils/loadKeyPairIdentities ^ClientSession session
+                                                           resource-key
+                                                           (ByteArrayInputStream. (.getBytes ^String tunnel-private-key "UTF-8"))
```

Comment:
shouldn't this input stream be used with `with-open`?

Reply:
I guess since it's a `ByteArrayInputStream` `.close()` is a no-op but IMO it's better just to always use `with-open` instead of having to remember which Input/Output stream types don't *really* need it

Reply:
yeah, it wouldn't hurt